### PR TITLE
Make crafting practice occur every turn instead of occasionally.

### PR DESCRIPTION
**Pull request by kevingranade**
#### Summary
None

#### Purpose of change
Addresses #47006 
Addresses #47172 
Practice from crafting was occurring very periodically in big chunks, causing confusion.

#### Describe the solution
Made practice ticks continuous instead, nominally occurring every turn unless the player's crafting speed is impaired, which scales down practice rate.
The end result isn't very different, it instead drops to 1 in about 15 minutes or so and then stays there for the duration of the craft.

#### Describe alternatives you've considered
Overhaul the focus pool and focus expenditure to make spending focus a very long term activity. This is blocked behind introducing practice actions for combat skills though, because once the super long duration practice mechanism is in place, gaining skill from grinding combat will be an extraordinarily slow process.
I'm not totally sure about the change to focus depletion, the old system had both good and bad aspects.

#### Testing
Ran and updated tests, the numbers are more consistent than they used to be, I think I fixed some bugs in the process of this PR.
Spawned a character, crafted some stuff, and watched focus while repeatedly halting crafting.  As expected focus dropped precipitously and then stayed down regardless of what I crafted.

#### Additional context
The extremely rapid depletion of focus is expected but not something we want to keep long term.
The original model for focus was that you lose 1% of your focus per practice action, and the rate of practice is throttled by low focus.  This results in something like 250 turns to deplete your focus fully, which is far too fast.

### CODE REVIEW

**Comment by PatrikLundell** [comment]
Different activities should probably both train skills and deplete focus at different rates. Making checks every turn makes sense, but the granularity ought to be finer so that you can lose e.g. 0.005% focus and have a chance (less than 100%) to gain e.g. 0.001% per tick over a 10000 second (roughly 3 hours) reading session.

If focus loss is kept at a single percent granularity it should have a probability to be lost. If that approach is made, I'd consider to couple focus loss to skill gain so that the cases where focus is lost are the cases where skill is gained (or has a chance to be gained). I'd prefer the first option, though.

File: src/crafting.cpp [hunk 1]
```diff
@@ -859,7 +859,7 @@ void Character::start_craft( craft_command &command, const cata::optional<tripoi
         craft.tname() );
 }
 
-void Character::craft_skill_gain( const item &craft, const int &multiplier )
+void Character::craft_skill_gain( const item &craft, const int &num_practice_ticks )
```
Comment (BrettDong) [inline src/crafting.cpp:862]: Parameters in declaration needs to be renamed as well.

**Comment by kevingranade** [comment]
> Different activities should probably both train skills and deplete focus at different rates.

I completely disagree, we're moving toward things being more uniform, currently the code is a mess of different skills having wildly divergent logic for how quickly they level, and there's no rationale for it, if you want "gains" to be faster, you can adjust what the impact of gaining a level of a skill means without tweaking how quickly practice happens.

> Making checks every turn makes sense, but the granularity ought to be finer so that you can lose e.g. 0.005% focus and have a chance (less than 100%) to gain e.g. 0.001% per tick over a 10000 second (roughly 3 hours) reading session.

Why?

> If focus loss is kept at a single percent granularity it should have a probability to be lost.

I'm moving everything possible away from the "chance to happen per tick" model.  It's a very attractive concept and I've used it extensively in the past, but once you start adding in testing it turns out that we are not performing enough iterations of these actions to make the "law of large numbers" kick in, so instead of being a nice way to handle sub-unit probabilities, it instead injects a significant degree of randomness into the system.

> If that approach is made, I'd consider to couple focus loss to skill gain so that the cases where focus is lost are the cases where skill is gained (or has a chance to be gained). I'd prefer the first option, though.

I think you're misunderstanding something, it no longer reduces the focus pool by 1% per practice action, and it now does tie focus loss to skill gain. The note at the end is there to sketch out what the status quo is (focus pool drains really fast) and why.  It's also not something I'm prepared to tackle right now, I know what I want to happen in broad strokes, but it is not a critical issue that needs addressing right now the way the crafting breakage (and the fallout from fixing the crafting breakage) was, so I'm deferring it till later.

**Comment by actual-nh** [comment]
> I think you're misunderstanding something, it no longer reduces the focus pool by 1% per practice action, and it now does tie focus loss to skill gain. The note at the end is there to sketch out what the status quo is (focus pool drains really fast) and why. It's also not something I'm prepared to tackle right now, I know what I want to happen in broad strokes, but it is not a critical issue that needs addressing right now the way the crafting breakage (and the fallout from fixing the crafting breakage) was, so I'm deferring it till later.

What is the desired curve in terms of focus loss?

**Comment by kevingranade** [comment]
Further testing showed that I hadn't captured the focus decay curve properly, as it wasn't something I really intended to mess with.  The result was that focus plummeted all the way to 1 or so and stayed there.  Then I tested with melee skill practice, which as it turns out I had COMPLETELY broken.

So basically I'm not touching character::practice() anymore and restricting my changes to just the crafting skill gain code and tests thereof.

The status quo is now back to "you expend 1% of your current focus every time you practice anything".

I'll have to circle back around when I feel like overhauling focus, the future plan is to get rid of all this reactive stuff and just make it a really big pool that regenerates and gets spent fairly slowly, something like several hours of practice to drain it fully.  Hopefully these new tests will help out once I decide to tackle that.

File: src/activity_actor.cpp [hunk 2]
```diff
@@ -1754,10 +1754,20 @@ void craft_activity_actor::do_turn( player_activity &act, Character &crafter )
     // This is to ensure we don't over count skill steps
     craft.item_counter = std::min( craft.item_counter, 10'000'000 );
 
-    // Skill and tools are gained/consumed after every 5% progress
+    // This nominal craft time is also how many practice ticks to perform
+    // spread out evenly across the actual duration.
+    const float total_practice_ticks = rec.time_to_craft_moves( crafter,
```
Comment (BrettDong) [inline src/activity_actor.cpp:1759]: rvalue here and the division next line are both of `double` type.
```suggestion
    const double total_practice_ticks = rec.time_to_craft_moves( crafter,
```

File: src/activity_actor.cpp [hunk 3]
```diff
@@ -1754,10 +1754,20 @@ void craft_activity_actor::do_turn( player_activity &act, Character &crafter )
...
+                                       recipe_time_flag::ignore_proficiencies ) / 100.0;
```
Comment (kevingranade) [inline src/activity_actor.cpp:1760]:
```suggestion
    const double total_practice_ticks = rec.time_to_craft_moves( crafter,
                                        recipe_time_flag::ignore_proficiencies ) / 100.0;
```

**Comment by actual-nh** [comment]
> I'm moving everything possible away from the "chance to happen per tick" model. It's a very attractive concept and I've used it extensively in the past, but once you start adding in testing it turns out that we are not performing enough iterations of these actions to make the "law of large numbers" kick in, so instead of being a nice way to handle sub-unit probabilities, it instead injects a significant degree of randomness into the system.

If there are instances when "chance to happen per tick" looks to be, aside from the over-dispersion,  the best model, I can see a few ways to regulate this. Most are based around having something be sampling without replacement instead of sampling with replacement - one can also look at this as giving you independent control over both mean and variance/SD. See #47218 for more.

**Pull request merged**